gdal/ogr/ogrsf_frmts/ngw/ogrngwlayer.cpp

```diff
@@ -0,0 +1,1116 @@
+    FillMetadata( oResourceJsonObject );
+
+    SetDescription( poFeatureDefn->GetName() );
+}
+
+/*
+ * OGRNGWLayer()
+ */
+OGRNGWLayer::OGRNGWLayer( OGRNGWDataset *poDSIn, const std::string &osNameIn,
+    OGRSpatialReference *poSpatialRef, OGRwkbGeometryType eGType,
+    const std::string &osKeyIn, const std::string &osDescIn ) :
+    osResourceId("-1"),
+    poDS(poDSIn),
+    bFetchedPermissions(false),
+    nFeatureCount(0),
+    oNextPos(moFeatures.begin()),
+    nPageStart(0),
+    bNeedSyncData(false),
+    bNeedSyncStructure(false)
+{
+    poFeatureDefn = new OGRFeatureDefn( osNameIn.c_str() );
+    poFeatureDefn->Reference();
+
+    poFeatureDefn->SetGeomType( eGType );
+
+    if( poSpatialRef )
+    {
+        if( poFeatureDefn->GetGeomFieldCount() != 0 )
+        {
+            poFeatureDefn->GetGeomFieldDefn(0)->SetSpatialRef( poSpatialRef );
+        }
+    }
+
+    if( !osDescIn.empty() )
+    {
+        OGRLayer::SetMetadataItem( "description", osDescIn.c_str() );
+    }
+    if( !osKeyIn.empty() )
+    {
+        OGRLayer::SetMetadataItem( "keyname", osKeyIn.c_str() );
+    }
+
+    SetDescription( poFeatureDefn->GetName() );
+}
+
+/*
+ * ~OGRNGWLayer()
+ */
+OGRNGWLayer::~OGRNGWLayer()
+{
+    if( !soChangedIds.empty() )
+    {
+        bNeedSyncData = true;
+    }
+    SyncFeatures();
+    FreeFeaturesCache();
+    if( poFeatureDefn != nullptr )
+    {
+        poFeatureDefn->Release();
+    }
+}
+
+/*
+ * FreeFeaturesCache()
+ */
+void OGRNGWLayer::FreeFeaturesCache()
+{
+    for (std::map<GIntBig, OGRFeature*>::iterator it = moFeatures.begin();
+        it != moFeatures.end(); ++it)
+    {
+        OGRFeature::DestroyFeature( it->second );
+    }
+    moFeatures.clear();
+}
+
+/*
+ * GetResourceId()
+ */
+std::string OGRNGWLayer::GetResourceId() const
+{
+    return osResourceId;
+}
+
+/*
+ * Delete()
+ */
+bool OGRNGWLayer::Delete()
+{
+    if( osResourceId == "-1")
+    {
+        return true;
+    }
+
+    // Headers free in DeleteResource method.
+    return NGWAPI::DeleteResource(poDS->GetUrl(), osResourceId, poDS->GetHeaders());
+}
+
+/*
+ * ResetReading()
+ */
+void OGRNGWLayer::ResetReading()
+{
+    SyncToDisk();
+    if( poDS->GetPageSize() != -1 )
+    {
+        FreeFeaturesCache();
+        nPageStart = 0;
+    }
+    oNextPos = moFeatures.begin();
+}
+
+/*
+ * SetNextByIndex()
+ */
+OGRErr OGRNGWLayer::SetNextByIndex( GIntBig nIndex )
+{
+    SyncToDisk();
+    if( nIndex < 0)
+    {
+        CPLError(CE_Failure, CPLE_AppDefined,
+            "Feature index must be greater or equal 0. Got " CPL_FRMT_GIB, nIndex);
+        return OGRERR_FAILURE;
+    }
+    if( poDS->GetPageSize() != -1 )
+    {
+        if( nPageStart < nIndex && nIndex >= nPageStart - poDS->GetPageSize() )
+        {
+            oNextPos = moFeatures.begin();
+            std::advance(oNextPos, nIndex);
+        }
+        else
+        {
+            nPageStart = nIndex / poDS->GetPageSize() * poDS->GetPageSize();
```

Comment:
is it guaranteed that GetPageSize() != 0 ?

Reply:
Fix per 4579aeab927d70ddab6142340bf5550c6789b42e